modules/reitit-dev/src/reitit/dev/pretty.cljc

```diff
@@ -0,0 +1,334 @@
+
+  (visit-meta [this m x]
+    (if print-meta
+      [:align [:span "^" (fipp.visit/visit this m)] :line (fipp.visit/visit* this x)]
+      (fipp.visit/visit* this x)))
+
+  (visit-var [_ x]
+    [:text (str x)])
+
+  (visit-pattern [_ x]
+    [:text (pr-str x)])
+
+  (visit-record [this x]
+    (fipp.visit/visit this (fipp.ednize/record->tagged x))))
+
+(defn printer
+  ([]
+   (printer nil))
+  ([options]
+   (map->EdnPrinter
+     (merge
+       {:width 80
+        :symbols {}
+        :print-length *print-length*
+        :print-level *print-level*
+        :print-meta *print-meta*}
+       options))))
+
+(defn pprint
+  ([x] (pprint x {}))
+  ([x options]
+   (let [printer (printer (dissoc options :margin))
+         margin (apply str (take (:margin options 0) (repeat " ")))]
+     (binding [*print-meta* false]
+       (fipp.engine/pprint-document [:group margin [:group (fipp.visit/visit printer x)]] options)))))
+
+(defn print-doc [doc printer]
+  (fipp.engine/pprint-document doc {:width (:width printer)}))
+
+(defn repeat-str [s n]
+  (apply str (take n (repeat s))))
+
+;; TODO: this is hack, but seems to work and is safe.
+(defn source-str [[target _ file line]]
+  (try
+    (if (and (not= 1 line))
+      (let [file-name (str/replace file #"(.*?)\.\S[^\.]+" "$1")
+            target-name (name target)
+            ns (str (subs target-name 0 (str/index-of target-name (str "user" "$"))) file-name)]
+        (str ns ":" line))
+      "repl")
+    (catch #?(:clj Exception, :cljs js/Error) _
+      "unknown")))
+
+(defn title [message source {:keys [width]}]
+  (let [between (- width (count message) 8 (count source))]
+    [:group
+     (color :title-dark "-- ")
+     (color :title message " ")
+     (color :title-dark (repeat-str "-" between) " ")
+     (color :title source) " "
+     (color :title-dark (str "--"))]))
+
+(defn footer [{:keys [width]}]
+  (color :title-dark (repeat-str "-" width)))
+
+(defn text [& text]
+  (apply color :text text))
+
+(defn edn
+  ([x] (edn x {}))
+  ([x options]
+   (with-out-str (pprint x options))))
+
+(defn exception-str [message source printer]
+  (with-out-str
+    (print-doc
+      [:group
+       (title "Router creation failed" source printer)
+       [:break] [:break]
+       message
+       [:break]
+       (footer printer)]
+      printer)))
+
+(defmulti format-exception (fn [type _ _] type))
+
+(defn exception [e]
+  (let [data (-> e ex-data :data)
+        message (format-exception (-> e ex-data :type) #?(:clj (.getMessage ^Exception e) :cljs (ex-message e)) data)
+        source #?(:clj (->> e Throwable->map :trace
+                            (drop-while #(not= (name (first %)) "reitit.core$router"))
+                            (drop-while #(= (name (first %)) "reitit.core$router"))
+                            next first source-str)
+                  :cljs "unkwown")]
+    (ex-info (exception-str message source (printer)) (assoc (or data {}) ::exception/cause e))))
```

Comment:
One more thing: you could've addedi a comment here about why we use custom `::exception/cause` here instead of the usual way handling it as the third argument of `ex-info`.